File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max_diameter = 0;   // public data member

    int dfs(TreeNode *node){
            if(node == NULL){
                return 0;
            }

            int left_height = dfs(node->left);
            int right_height = dfs(node->right);

            // Update the max_diameter
            max_diameter = max(max_diameter, left_height + right_height);

            // Return the height of current node
            return max(left_height, right_height) + 1;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return max_diameter;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0

        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            
            left_height = dfs(node.left)
            right_height = dfs(node.right)

            # Update max_diameter:
            self.max_diameter = max(self.max_diameter, left_height + right_height)

            # Return the height of current node
            return max(left_height, right_height) + 1
        
        dfs(root)
        return self.max_diameter
